### PR TITLE
adding torch.nn.ConvTranspose2d

### PR DESCRIPTION
**Pull request by matthieutrs**
Thanks for this very nice repo! I needed to convert models containing `torch.nn.ConvTranspose2d` but this is not completely straightforward as torch and lax do not perform the same transposed conv. 

This PR is essentially a merging of [https://github.com/google/jax/pull/5772](https://github.com/google/jax/pull/5772) to solve this issue. 

Atm it relies on numpy; if this is a problem I think this could be avoided but I didn't have time to remove the dependency yet.

### CODE REVIEW

File: tests/test_all_the_things.py [hunk 1]
```diff
@@ -382,3 +382,19 @@ def test_vit_b16():
...
+  input_batch = random.normal(random.PRNGKey(123), (3, 8, 16, 16))
+  res_torch = model(j2t(input_batch))
+
+  jaxified_module = t2j(model)
+  res_jax = jaxified_module(input_batch, state_dict=parameters)
+  res_jax_jit = jit(jaxified_module)(input_batch, state_dict=parameters)
+
+  # Test forward pass with and without jax.jit
+  aac(res_jax, res_torch.numpy(force=True), atol=1e-1)
+  aac(res_jax_jit, res_torch.numpy(force=True), atol=1e-1)
```
Comment (samuela) [inline tests/test_all_the_things.py:400]: A more exhaustive test suite:
```suggestion
def test_conv_transpose2d():
  for in_channels in [2, 4, 8]:
    for out_channels in [2, 4, 8]:
      for kernel_size in [(1, 1), (2, 2), (3, 3), 1, 2, 3, (1, 2), (2, 3)]:
        for stride in [(1, 1), (2, 2), (3, 3), 1, 2, 3, (1, 2), (2, 3)]:
          for bias in [False, True]:
            print(in_channels, out_channels, kernel_size, stride, bias)
            model = torch.nn.ConvTranspose2d(in_channels, out_channels, kernel_size, stride, bias=bias)

            parameters = {k: t2j(v) for k, v in model.named_parameters()}
            input_batch = random.normal(random.PRNGKey(123), (3, in_channels, 16, 16))
            res_torch = model(j2t(input_batch))

            jaxified_module = t2j(model)
            res_jax = jaxified_module(input_batch, state_dict=parameters)

            print(res_jax.shape, res_torch.numpy(force=True).shape)
            aac(res_jax, res_torch.numpy(force=True), atol=1e-1)
```
reveals a failure with `stride != 1`. For example, `in_channels=2, out_channels=2, kernel_size=(1, 1), stride=(2, 2), bias=False`. The PyTorch implementation gives an output of shape `(3, 2, 31, 31)` whereas the JAX version gives an output of shape `(3, 2, 32, 32)`. Any idea what might cause this? A quick glance didn't reveal anything to me.

As a workaround, we could support only `stride == 1` for the time being. But we should make that explicit as opposed to a silent failure case.

File: torch2jax/__init__.py [hunk 2]
```diff
@@ -1,13 +1,25 @@
 import copy
 import functools
 import math
-from typing import Optional
+from typing import (Any, Optional, Sequence, Union, Tuple)
+
+import numpy as np
```
Comment (samuela) [inline torch2jax/__init__.py:6]: Is there any way we can avoid introducing a dependency on numpy? Why not use `jax.numpy` instead?

**Review by samuela — CHANGES_REQUESTED** [review]
Thanks for putting this together @matthieutrs! My main concern is that there seems to be a bug when `stride != 1`. I tried poking on it briefly, but couldn't figure it out... any idea what might be necessary here?

Would an implementation with `assert stride == 1` work for your usecase?

File: tests/test_all_the_things.py (hunk 1 above)
Comment (matthieutrs) [reply tests/test_all_the_things.py:400]: Thanks for raising this! 

Actually, the problem comes from that the jax implementation assumes a certain `output_padding` of `torch.nn.ConvTranspose2d`, which needs to be set to a nonzero value whenever (stride-kernel_size) becomes greater than 0. The test below passes:

```suggestion
def test_conv_transpose2d():
  for in_channels in [2, 4, 8]:
    for out_channels in [2, 4, 8]:
      for kernel_size in [(1, 1), (2, 2), (3, 3), 1, 2, 3, (1, 2), (2, 3)]:
        for stride in [(1, 1), (2, 2), (3, 3), 1, 2, 3, (1, 2), (2, 3)]:
          for bias in [False, True]:
            print(in_channels, out_channels, kernel_size, stride, bias)
            if isinstance(kernel_size, int):
               kernel_size = (kernel_size, kernel_size)
            if isinstance(stride, int):
               stride = (stride, stride)
            output_padding = (max(stride[0]-kernel_size[0], 0), max(stride[1]-kernel_size[1], 0))
            model = torch.nn.ConvTranspose2d(in_channels, out_channels, kernel_size, stride, bias=bias, output_padding=output_padding)
            parameters = {k: t2j(v) for k, v in model.named_parameters()}
            input_batch = random.normal(random.PRNGKey(123), (3, in_channels, 16, 16))
            res_torch = model(j2t(input_batch))
            jaxified_module = t2j(model)
            res_jax = jaxified_module(input_batch, state_dict=parameters)
            print(res_jax.shape, res_torch.numpy(force=True).shape)
            aac(res_jax, res_torch.numpy(force=True), atol=1e-1)
```

Adding an assertion in `def convtranspose2d` should allow to check for that.

**Comment by matthieutrs** [comment]
Thanks a lot for the careful review!
1. The dependency to numpy has been removed;
2. The failling tests for some strides/kernel sizes was due to jax assuming a certain `output_padding` in `torch.nn.ConvTranspose2d`, I've added an assertion in the def of `conv_transpose2d` (see above);

Let me know if there are other things that need to be updated!

File: torch2jax/__init__.py [hunk 3]
```diff
@@ -230,6 +230,247 @@ def conv2d(input, weight, bias=None, stride=1, padding=0, dilation=1, groups=1):
...
+    lhs=coerce(input),
+    rhs=coerce(weight),
+    strides=stride,
+    padding="VALID",
+    dimension_numbers=("NCHW", "OIHW", "NCHW"),
+    dilation=dilation,
+  )
+  if bias is not None:
+    res += coerce(bias)[jnp.newaxis, :, jnp.newaxis, jnp.newaxis]
+  return res
```
Comment (samuela) [inline torch2jax/__init__.py:256]:
```suggestion
@implements(torch.nn.functional.conv_transpose2d)
def conv_transpose2d(input, weight, bias=None, stride=1, padding=0, output_padding=0, groups=1, dilation=1):
  # This implementation is taken from this PR https://github.com/google/jax/pull/5772
  assert input.ndim == 4, "TODO: implement non-batched input"
  assert groups == 1, "TODO: implement groups != 1"

  ph, pw = (padding, padding) if isinstance(padding, int) else padding
  res = gradient_based_conv_transpose(
    lhs=coerce(input),
    rhs=coerce(weight),
    strides=stride,
    padding=[(ph, ph), (pw, pw)],
    output_padding=output_padding,
    dilation=dilation,
    dimension_numbers=("NCHW", "OIHW", "NCHW"),
  )
  if bias is not None:
    res += coerce(bias)[jnp.newaxis, :, jnp.newaxis, jnp.newaxis]
  return res
```

File: tests/test_all_the_things.py [hunk 4]
```diff
@@ -438,3 +438,25 @@ def test_vit_b16():
...
+            output_padding = (max(stride[0] - kernel_size[0], 0), max(stride[1] - kernel_size[1], 0))
+            model = torch.nn.ConvTranspose2d(
+              in_channels, out_channels, kernel_size, stride, bias=bias, output_padding=output_padding
+            )
+            parameters = {k: t2j(v) for k, v in model.named_parameters()}
+            input_batch = random.normal(random.PRNGKey(123), (3, in_channels, 16, 16))
+            res_torch = model(j2t(input_batch))
+            jaxified_module = t2j(model)
+            res_jax = jaxified_module(input_batch, state_dict=parameters)
+            aac(res_jax, res_torch.numpy(force=True), atol=1e-1)
```
Comment (samuela) [inline tests/test_all_the_things.py:462]:
```suggestion
def test_conv_transpose2d():
  for in_channels in [1, 2]:
    for out_channels in [1, 2]:
      for kernel_size in [1, 2, (1, 2)]:
        for stride in [1, 2, (1, 2)]:
          for padding in [(0, 0), 1, 2, (1, 2)]:
            for output_padding in [0, 1, 2, (1, 2)]:
              for bias in [False, True]:
                for dilation in [1, 2, (1, 2)]:
                  model = torch.nn.ConvTranspose2d(
                    in_channels,
                    out_channels,
                    kernel_size,
                    stride=stride,
                    padding=padding,
                    output_padding=output_padding,
                    bias=bias,
                    dilation=dilation,
                  )
                  params = {k: random.normal(random.PRNGKey(123), v.shape) for k, v in model.named_parameters()}
                  model.load_state_dict({k: j2t(v) for k, v in params.items()})

                  input_batch = random.normal(random.PRNGKey(123), (3, in_channels, 16, 16))
                  try:
                    res_torch = model(j2t(input_batch))
                  except RuntimeError:
                    # RuntimeError: output padding must be smaller than either stride or dilation
                    continue

                  res_jax = t2j(model)(input_batch, state_dict=params)
                  aac(res_jax, res_torch.numpy(force=True), atol=1e-4)
```

**Review by samuela — CHANGES_REQUESTED** [review]
Figured out a fix such that it works for basically all configuration options to `torch.nn.ConvTranspose2d`! The one exception being `groups != 1`, but I'm not worried about that for now.

Additionally,
* Make the test case deterministic by skirting reliance on PyTorch randomness for weights
* Decrease `atol`

**Comment by matthieutrs** [comment]
Thanks a lot! It indeed works on my architectures.

**Review by samuela — APPROVED** [review]

**Comment by samuela** [comment]
Thanks so much @matthieutrs! You get a special prize for being the first person to merge a PR on torch2jax! 🌟

**Pull request merged**